lib/lightstep/tracer/tracer.rb

```diff
@@ -0,0 +1,374 @@
+      @dropped_logs = Concurrent::AtomicFixnum.new
+      @dropped_spans = Concurrent::AtomicFixnum.new
+
+      start_time = now_micros.to_i
+      @guid = generate_guid
+      @tracer_report_start_time = start_time
+      @tracer_last_flush_micros = start_time
+
+      @tracer_thrift_runtime = {
+        'guid' => guid,
+        'start_micros' => start_time,
+        'group_name' => component_name,
+        'attrs' => [
+          {"Key" => "lightstep_tracer_platform", "Value" => "ruby"},
+          {"Key" => "lightstep_tracer_version",  "Value" => LightStep::Tracer::VERSION},
+          {"Key" => "ruby_version",              "Value" => RUBY_VERSION}
+        ]
+      }.freeze
+
+      if !transport.nil?
+        if !(LightStep::Transport::Base === transport)
+          raise ConfigurationError, "transport is not a LightStep transport class: #{transport}"
+        end
+        @tracer_transport = transport
+      else
+        if access_token.nil?
+          raise ConfigurationError, "you must provide an access token or a transport"
+        end
+        @tracer_transport = Transport::HTTPJSON.new(access_token: access_token)
+      end
+
+      # At exit, flush this objects data to the transport and close the transport
+      # (which in turn will send the flushed data over the network).
+      at_exit do
+        flush
+        @tracer_transport.close
+      end
+    end
+
+    def push_with_max(arr, item, max)
+      max = 1 if max < 1
+      arr.push item
+
+      if arr.size <= max
+        return false
+      end
+
+      while arr.size > max
+        arr.delete_at(rand(arr.size))
+      end
+      return true
+    end
+
+    def flush_if_needed
+      return unless enabled?
+
+      delta = now_micros - @tracer_last_flush_micros
+
+      # Set a bound on maximum flush frequency
+      return if delta < min_flush_period_micros
+
+      # Look for a trigger that a flush is warranted
+      # Set a bound of minimum flush frequency
+      if delta > max_flush_period_micros ||
+         @tracer_log_records.length >= max_log_records / 2 ||
+         @tracer_span_records.length >= max_span_records / 2
+        flush
+      end
+    end
+
+    def now_micros
+      micros(Time.now)
+    end
+
+    private
+
+    def micros(time)
+      (time.to_f * 1E6).floor
+    end
+
+    def inject_to_text_map(span, carrier)
+      carrier[CARRIER_TRACER_STATE_PREFIX + 'spanid'] = span.guid
+      carrier[CARRIER_TRACER_STATE_PREFIX + 'traceid'] = span.trace_guid unless span.trace_guid.nil?
+      carrier[CARRIER_TRACER_STATE_PREFIX + 'sampled'] = 'true'
+
+      span.baggage.each do |key, value|
+        carrier[CARRIER_BAGGAGE_PREFIX + key] = value
+      end
+    end
+
+    def extract_from_text_map(operation_name, carrier)
+      span = Span.new(
+        tracer: self,
+        operation_name: operation_name,
+        start_micros: now_micros,
+        child_of_guid: carrier[CARRIER_TRACER_STATE_PREFIX + 'spanid'],
```

Comment:
I didn't modify this part of the code, I just assumed it worked. I mainly just moved it around. I am fairly sure Rack normalizes case already (or provides indifferent access) but I can look into it.